### DISCUSSION TITLE
How to ignore Gravity with Dynamic RigidBody?

### DISCUSSION THREAD

**Discussion by TarekAS** [jcornaz/heron#293] [Q&A] [answered]
Hi,

I'm making a flying camera that needs to have collision on and not be affected by gravity. Since the `RigidBody::KinematicPositionBased` does not collide, I'm using a `RigidBody::Dynamic` instead. However, that enables gravity on the object obviously.

I tried:
- Setting density to 0.0 (causing expected crash on collision as it needs to be > 0)
- Setting density to 0.0000000001 (changes nothing, worth a try)
- Finding a way to set mass to 0
- Finding a way to make `KinematicPositionBased` collide.

I haven't tried:
- Using a different physics layer with 0 gravity for this object only.

Any advice?

**Comment by jcornaz** [comment, marked as answer]
HI, 

> * Setting density to 0.0 (causing expected crash on collision as it needs to be > 0)
> * Setting density to 0.0000000001 (changes nothing, worth a try)
> * Finding a way to set mass to 0

Like in the real world, gravitational acceleration is not impacted by the mass of the object. Therefore, changing the mass (or density) is not supposed to be of any help.

> Finding a way to make KinematicPositionBased collide.

By design, kinematic bodies are not supposed to be affected by forces. The "impulse" generated by contact-resolution is one such force that is not supposed to affect kinematic bodies, *by design*.

---

As you said, `Dynamic` objects are supposed to be affected by forces, and gravity is one of them. Though you may counteract the acceleration by addding an [Acceleration](https://docs.rs/heron/latest/heron/struct.Acceleration.html) to your object that is the inverse of your world gravity.

**Reply by TarekAS** [reply]
Thanks for the reply, and the quick physics lesson! (forgot that gravity bends light 😄)

I managed to fake removing gravity by re-initializing the rigidbody component on each frame:
```rust
fn fake_remove_gravity_system(mut q: Query<(&FlyCam, &mut RigidBody)>) {
    let (_, mut rb) = q.single_mut();
    *rb = RigidBody::Dynamic
}
```
This allows you to keep collision and pseudo-remove gravity.

Obviously this is a hack, but I only need it for debugging purposes anyway.
The proper way is a you suggested, applying acceleration inverse of the world gravity.

**Reply by TarekAS** [reply]
The inverse acceleration method:

```rust
<your_entity>
    .insert(RigidBody::Dynamic)
    .insert(Acceleration::from_linear(Vec3::Y * 9.81))
```
